code/menuui/optionsmenu.cpp

```diff
@@ -1374,12 +1398,21 @@ void options_detail_sliders_update()
 	Detail.nebula_detail = Detail_sliders[gr_screen.res][NEBULA_DETAIL_SLIDER].slider.pos;
 	neb2_set_detail_level(Detail.nebula_detail);
 
-	Detail.hardware_textures = Detail_sliders[gr_screen.res][HARDWARE_TEXTURES_SLIDER].slider.pos;	
+	Detail.hardware_textures = Detail_sliders[gr_screen.res][HARDWARE_TEXTURES_SLIDER].slider.pos;
 	Detail.num_small_debris = Detail_sliders[gr_screen.res][SHARD_CULLING_SLIDER].slider.pos;
 	Detail.shield_effects = Detail_sliders[gr_screen.res][SHIELD_DETAIL_SLIDER].slider.pos;
 	Detail.num_stars = Detail_sliders[gr_screen.res][NUM_STARS_SLIDER].slider.pos;
 	Detail.num_particles = Detail_sliders[gr_screen.res][NUM_PARTICLES_SLIDER].slider.pos;
-	Detail.lighting = Detail_sliders[gr_screen.res][LIGHTING_SLIDER].slider.pos;
+
+	// If the new lighting setting is above 3 and the old one was below or the reverse,
+	// we need to recompile all shaders we have to account for the changed lighting model.
+	
+	if (Detail.lighting != Detail_sliders[gr_screen.res][LIGHTING_SLIDER].slider.pos) {
+		Detail.lighting = Detail_sliders[gr_screen.res][LIGHTING_SLIDER].slider.pos;
+		shader_compile_status = false;
+		shader_compile_started = false;
+		popup_till_condition(recompile_shaders, POPUP_CANCEL, "Recompiling shaders");
```

Comment:
This doesn't actually display a popup with changing text but since that would likely require adding a new popup type it's not really in the scope of these changes.

Reply:
It does, although indirectly. The shader recompilation callback constructs a status string, which is then applied on-frame via recompile_shaders() calling popup_change_text.
It's a clunky solution, but in the absence of a new popup type specifically for situations like these, the best available.

Reply:
`gr_recompile_shaders` doesn't return until all the shaders are compiled so effectively `recompile_shaders` is only called once and the display never updates (at least when I tested the changes it didn't).